vehicle/subaru/identity.go

```diff
@@ -0,0 +1,201 @@
+	uri := fmt.Sprintf("%s/%s", BaseUrl, AuthenticationPath)
+
+	// Update callbacks with credentials
+	for id, cb := range auth.Callbacks {
+		switch cb.Type {
+		case "NameCallback":
+			outputValue, ok := cb.Output[0].Value.(string)
+			if ok && outputValue == "User Name" {
+				auth.Callbacks[id].Input[0].Value = user
+			}
+		case "PasswordCallback":
+			auth.Callbacks[id].Input[0].Value = password
+			passwordSet = true
+		}
+	}
+
+	// Send authentication request
+	req, err := request.New(http.MethodPost, uri, request.MarshalJSON(auth), request.JSONEncoding)
+	if err != nil {
+		return nil, err
+	}
+
+	// If we've already set the password, expect a token response
+	if passwordSet {
+		var token Token
+		if err := v.DoJSON(req, &token); err != nil {
+			return nil, err
+		}
+		return &token, nil
+	}
+
+	// Otherwise continue with Auth flow
+	var res Auth
+	if err := v.DoJSON(req, &res); err != nil {
+		return nil, err
+	}
+
+	// Continue authentication flow
+	return v.authenticate(res, user, password, passwordSet)
+}
+
+func (v *Identity) authorize(token Token) (string, error) {
+	uri := fmt.Sprintf("%s/%s", BaseUrl, AuthorizationPath)
+	req, err := request.New(http.MethodGet, uri, nil, map[string]string{
+		"Cookie": fmt.Sprintf("iPlanetDirectoryPro=%s", token.TokenID),
+	})
+	if err != nil {
+		return "", err
+	}
+	var param request.InterceptResult
+	v.Client.CheckRedirect, param = request.InterceptRedirect("code", true)
+	var code string
+	if _, err = v.Do(req); err == nil {
+		code, err = param()
+	}
+	return code, err
+}
+
+func (v *Identity) fetchTokenCredentials(code string) error {
+	uri := fmt.Sprintf("%s/%s", BaseUrl, AccessTokenPath)
+
+	data := url.Values{
+		"client_id":     {ClientID},
+		"code":          {code},
+		"redirect_uri":  {RedirectURI},
+		"grant_type":    {"authorization_code"},
+		"code_verifier": {"plain"},
+	}
+
+	headers := request.URLEncoding
+	headers["Authorization"] = AppAuthorization
+	req, err := request.New(http.MethodPost, uri, strings.NewReader(data.Encode()), headers)
+	if err != nil {
+		return err
+	}
+
+	var res struct {
+		oauth2.Token
+		IDToken string `json:"id_token"`
+	}
+	if err = v.DoJSON(req, &res); err != nil {
+		return fmt.Errorf("failed to fetch token credentials: %w", err)
+	}
+
+	// Parse ID token without verification to extract UUID
+	parser := jwt.NewParser(jwt.WithoutClaimsValidation())
+	token, _, err := parser.ParseUnverified(res.IDToken, jwt.MapClaims{})
+	if err != nil {
+		return fmt.Errorf("failed to parse id token: %w", err)
+	}
+
+	claims, ok := token.Claims.(jwt.MapClaims)
+	if !ok {
+		return fmt.Errorf("invalid token claims format")
+	}
+
+	uuid, ok := claims["uuid"].(string)
+	if !ok {
+		return fmt.Errorf("uuid claim missing or invalid in token")
+	}
+
+	v.uuid = uuid
+	v.TokenSource = oauth.RefreshTokenSource(util.TokenWithExpiry(&res.Token), v)
+	return nil
+}
+
+func (v *Identity) RefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
```

Comment:
this looks a lot like normal OAuth flow. If possible it would be nice to use an `oauth2.Config` and associated functions. The `headers` could be set using a headers decorator on the transport.